File: algorithms/python3/725.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def splitListToParts(self, root, k):
        """
        :type root: ListNode
        :type k: int
        :rtype: List[ListNode]
        """
        l = self.getLength(root)
        ps = l // k
        parts = []
        for i in range(l%k):
            node, root = self.getK(root, ps+1)
            parts.append(node)
        for i in range(l%k, k):
            node, root = self.getK(root, ps)
            parts.append(node)
        return parts

    def getK(self, root, k):
        if k == 0:
            return None, root
        p, q, k = root, root.next, k-1
        while k:
            p, q, k = q, q.next, k-1
        p.next = None
        return root, q

    def getLength(self, root):
        length = 0
        while root:
            length += 1
            root = root.next
        return length

File: algorithms/rust/725.rs
#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

impl Solution {
    #[allow(dead_code)]
    pub fn split_list_to_parts(
        mut head: Option<Box<ListNode>>,
        k: i32,
    ) -> Vec<Option<Box<ListNode>>> {
        let k = k as usize;
        let mut cnt = 0;
        let mut cur = &head;
        while let Some(node) = cur {
            cur = &node.next;
            cnt += 1;
        }

        let mut rslt = vec![];
        let mut part;

        for i in 0..k {
            let n = cnt / k + if i < cnt % k { 1 } else { 0 };
            let split = Self::take(head, n);
            part = split.0;
            head = split.1;
            rslt.push(part);
        }

        rslt
    }

    fn take(
        mut head: Option<Box<ListNode>>,
        mut n: usize,
    ) -> (Option<Box<ListNode>>, Option<Box<ListNode>>) {
        if head.is_none() {
            return (None, head);
        }

        let mut part = Box::new(ListNode::new(0));
        let mut tail = &mut part;
        while let Some(node) = head {
            tail.next = Some(node);
            tail = tail.next.as_mut().unwrap();
            head = tail.next.take();
            n -= 1;
            if n == 0 {
                break;
            }
        }
        (part.next, head)
    }
}

struct Solution;

fn main() {
    println!("Hello, world!");
}
